File: scripts/optimize_images.mjs
import { cpus } from "node:os";
import { extname, join, parse, resolve } from "node:path";
import { readdir, rename, rm, stat } from "node:fs/promises";
import sharp from "sharp";

const SUPPORTED_EXTENSIONS = new Set([".jpg", ".jpeg", ".png", ".webp", ".avif"]);
const SKIPPED_DIRECTORIES = new Set([".git", ".venv", "node_modules", "staticfiles"]);
const DEFAULT_ROOTS = ["main/static", "media", "products"];

function parsePositiveInteger(value, optionName) {
  const parsed = Number.parseInt(value, 10);
  if (!Number.isFinite(parsed) || parsed <= 0) {
    throw new Error(`${optionName} must be a positive integer.`);
  }
  return parsed;
}

function parseArguments(argv) {
  const options = {
    write: false,
    replace: false,
    format: "preserve",
    maxWidth: 2560,
    maxHeight: 2560,
    quality: 84,
    minSavingsPercent: 1,
    concurrency: Math.max(1, Math.min(4, cpus().length)),
    help: false,
    roots: [],
  };

  for (const argument of argv) {
    if (argument === "--help" || argument === "-h") options.help = true;
    else if (argument === "--write") options.write = true;
    else if (argument === "--audit") options.write = false;
    else if (argument === "--replace") options.replace = true;
    else if (argument.startsWith("--format=")) options.format = argument.split("=", 2)[1].toLowerCase();
    else if (argument.startsWith("--max-width=")) options.maxWidth = parsePositiveInteger(argument.split("=", 2)[1], "--max-width");
    else if (argument.startsWith("--max-height=")) options.maxHeight = parsePositiveInteger(argument.split("=", 2)[1], "--max-height");
    else if (argument.startsWith("--quality=")) options.quality = parsePositiveInteger(argument.split("=", 2)[1], "--quality");
    else if (argument.startsWith("--min-savings=")) options.minSavingsPercent = parsePositiveInteger(argument.split("=", 2)[1], "--min-savings");
    else if (argument.startsWith("--concurrency=")) options.concurrency = parsePositiveInteger(argument.split("=", 2)[1], "--concurrency");
    else if (argument.startsWith("--")) throw new Error(`Unknown option: ${argument}`);
    else options.roots.push(argument);
  }

  if (!new Set(["preserve", "webp", "avif"]).has(options.format)) {
    throw new Error("--format must be preserve, webp, or avif.");
  }
  if (options.quality > 100) throw new Error("--quality must not exceed 100.");
  if (options.replace && !options.write) throw new Error("--replace requires --write.");
  if (options.roots.length === 0) options.roots = DEFAULT_ROOTS;
  return options;
}

function printHelp() {
  console.log(`Usage: node scripts/optimize_images.mjs [paths...] [options]

Without --write the command performs a read-only audit.

Options:
  --write                 Write optimized files
  --format=preserve       Keep source formats (default)
  --format=webp|avif      Create the selected modern format
  --replace               Delete a source after successful conversion
  --max-width=2560        Maximum output width
  --max-height=2560       Maximum output height
  --quality=84            JPEG/WebP/AVIF quality, 1-100
  --min-savings=1         Skip files saving less than this percentage
  --concurrency=4         Maximum parallel image jobs
  --help                  Show this help`);
}

async function pathExists(path) {
  try {
    await stat(path);
    return true;
  } catch (error) {
    if (error.code === "ENOENT") return false;
    throw error;
  }
}

async function collectImages(root, output) {
  if (!(await pathExists(root))) return;
  const rootStat = await stat(root);
  if (rootStat.isFile()) {
    if (SUPPORTED_EXTENSIONS.has(extname(root).toLowerCase())) output.push(root);
    return;
  }

  const entries = await readdir(root, { withFileTypes: true });
  for (const entry of entries) {
    if (entry.isSymbolicLink()) continue;
    if (entry.isDirectory() && SKIPPED_DIRECTORIES.has(entry.name)) continue;
    const entryPath = join(root, entry.name);
    if (entry.isDirectory()) await collectImages(entryPath, output);
    else if (SUPPORTED_EXTENSIONS.has(extname(entry.name).toLowerCase())) output.push(entryPath);
  }
}

function formatBytes(bytes) {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 ** 2) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / 1024 ** 2).toFixed(2)} MB`;
}

function targetExtension(sourceExtension, format) {
  if (format === "preserve") return sourceExtension;
  return `.${format}`;
}

function configureEncoder(pipeline, extension, quality) {
  if (extension === ".jpg" || extension === ".jpeg") {
    return pipeline.jpeg({ quality, mozjpeg: true, chromaSubsampling: "4:4:4" });
  }
  if (extension === ".png") {
    return pipeline.png({ compressionLevel: 9, adaptiveFiltering: true });
  }
  if (extension === ".webp") {
    return pipeline.webp({ quality, alphaQuality: quality, effort: 6, smartSubsample: true });
  }
  if (extension === ".avif") {
    return pipeline.avif({ quality, effort: 6, chromaSubsampling: "4:4:4" });
  }
  throw new Error(`Unsupported output extension: ${extension}`);
}

async function inspectImage(filePath, options) {
  const fileStat = await stat(filePath);
  const metadata = await sharp(filePath, { failOn: "warning", limitInputPixels: false }).metadata();
  const width = metadata.autoOrient?.width ?? metadata.width ?? 0;
  const height = metadata.autoOrient?.height ?? metadata.height ?? 0;
  const needsResize = width > options.maxWidth || height > options.maxHeight;
  return { filePath, fileStat, metadata, width, height, needsResize };
}

async function optimizeImage(image, options) {
  const sourceExtension = extname(image.filePath).toLowerCase();
  const outputExtension = targetExtension(sourceExtension, options.format);
  const parsedPath = parse(image.filePath);
  const targetPath = join(parsedPath.dir, `${parsedPath.name}${outputExtension}`);
  const temporaryPath = join(parsedPath.dir, `.${parsedPath.name}.optimize-${process.pid}-${Date.now()}${outputExtension}`);

  let pipeline = sharp(image.filePath, { failOn: "warning", limitInputPixels: false }).rotate();
  if (image.needsResize) {
    pipeline = pipeline.resize({
      width: options.maxWidth,
      height: options.maxHeight,
      fit: "inside",
      withoutEnlargement: true,
      kernel: sharp.kernel.lanczos3,
    });
  }
  pipeline = configureEncoder(pipeline, outputExtension, options.quality).withMetadata({ orientation: undefined });

  try {
    const outputInfo = await pipeline.toFile(temporaryPath);
    const outputStat = await stat(temporaryPath);
    const savingsPercent = ((image.fileStat.size - outputStat.size) / image.fileStat.size) * 100;
    const worthWriting = image.needsResize || savingsPercent >= options.minSavingsPercent;

    if (!worthWriting) {
      await rm(temporaryPath, { force: true });
      return { status: "skipped", ...image, outputInfo, outputSize: outputStat.size, savingsPercent };
    }

    if (targetPath === image.filePath) {
      await rm(image.filePath, { force: true });
    } else {
      await rm(targetPath, { force: true });
    }
    await rename(temporaryPath, targetPath);
    if (options.replace && targetPath !== image.filePath) await rm(image.filePath, { force: true });

    return { status: "optimized", ...image, targetPath, outputInfo, outputSize: outputStat.size, savingsPercent };
  } catch (error) {
    await rm(temporaryPath, { force: true });
    throw error;
  }
}

async function runPool(items, concurrency, worker) {
  const results = new Array(items.length);
  let nextIndex = 0;
  async function runWorker() {
    while (nextIndex < items.length) {
      const currentIndex = nextIndex;
      nextIndex += 1;
      try {
        results[currentIndex] = await worker(items[currentIndex]);
      } catch (error) {
        results[currentIndex] = { status: "error", filePath: items[currentIndex], error };
      }
    }
  }
  await Promise.all(Array.from({ length: Math.min(concurrency, items.length) }, runWorker));
  return results;
}

async function main() {
  const options = parseArguments(process.argv.slice(2));
  if (options.help) {
    printHelp();
    return;
  }
  const images = [];
  for (const root of options.roots) await collectImages(resolve(root), images);
  const uniqueImages = [...new Set(images)].sort((left, right) => left.localeCompare(right));

  if (uniqueImages.length === 0) {
    console.log("No supported images found.");
    return;
  }

  if (options.format !== "preserve") {
    const outputOwners = new Map();
    for (const imagePath of uniqueImages) {
      const parsedPath = parse(imagePath);
      const outputPath = join(parsedPath.dir, `${parsedPath.name}.${options.format}`).toLowerCase();
      const existingOwner = outputOwners.get(outputPath);
      if (existingOwner && existingOwner !== imagePath) {
        throw new Error(`Output collision: ${existingOwner} and ${imagePath} both map to ${outputPath}.`);
      }
      outputOwners.set(outputPath, imagePath);
    }
  }

  const inspected = await runPool(uniqueImages, options.concurrency, (filePath) => inspectImage(filePath, options));
  const inspectionErrors = inspected.filter((item) => item.status === "error");
  const validImages = inspected.filter((item) => item.status !== "error");
  const totalBefore = validImages.reduce((sum, item) => sum + item.fileStat.size, 0);

  if (!options.write) {
    for (const image of validImages) {
      const marker = image.needsResize ? "RESIZE" : "OK";
      console.log(`${marker.padEnd(6)} ${image.width}x${image.height} ${formatBytes(image.fileStat.size).padStart(9)}  ${image.filePath}`);
    }
    console.log(`\nAudit: ${validImages.length} images, ${formatBytes(totalBefore)} total.`);
    console.log("No files changed. Add --write to optimize.");
  } else {
    const results = await runPool(validImages, options.concurrency, (image) => optimizeImage(image, options));
    let totalAfter = 0;
    let optimizedCount = 0;
    for (const result of results) {
      if (result.status === "optimized") {
        optimizedCount += 1;
        totalAfter += result.outputSize;
        console.log(`WRITE  ${formatBytes(result.fileStat.size)} -> ${formatBytes(result.outputSize)} (${result.savingsPercent.toFixed(1)}%)  ${result.targetPath}`);
      } else if (result.status === "skipped") {
        totalAfter += result.fileStat.size;
        console.log(`SKIP   ${result.filePath}`);
      } else {
        totalAfter += result.fileStat?.size ?? 0;
        console.error(`ERROR  ${result.filePath}: ${result.error.message}`);
      }
    }
    console.log(`\nOptimized ${optimizedCount}/${validImages.length} images.`);
    console.log(`Total: ${formatBytes(totalBefore)} -> ${formatBytes(totalAfter)} (${((1 - totalAfter / totalBefore) * 100).toFixed(1)}% smaller).`);
  }

  for (const failure of inspectionErrors) console.error(`ERROR  ${failure.filePath}: ${failure.error.message}`);
  if (inspectionErrors.length > 0) process.exitCode = 1;
}

main().catch((error) => {
  console.error(error.message);
  process.exitCode = 1;
});

File: main/management/commands/optimize_images.py
from pathlib import Path

from django.conf import settings
from django.core.management.base import BaseCommand, CommandError
from PIL import Image, ImageOps


SUPPORTED_EXTENSIONS = {".jpg", ".jpeg", ".png", ".webp"}
SKIPPED_DIRECTORIES = {".git", ".venv", "node_modules", "staticfiles", "__pycache__"}
RESPONSIVE_WIDTHS = (360, 640, 960, 1280, 1920)


class Command(BaseCommand):
    help = "Audit or optimize local project images with Pillow."

    def add_arguments(self, parser):
        parser.add_argument(
            "paths",
            nargs="*",
            help="Files or directories to scan. Defaults to MEDIA_ROOT and main/static.",
        )
        parser.add_argument("--write", action="store_true", help="Write optimized files.")
        parser.add_argument(
            "--format",
            choices=("preserve", "webp"),
            default="preserve",
            help="Output format. Default keeps source extension.",
        )
        parser.add_argument("--replace", action="store_true", help="Delete source after converting to another format.")
        parser.add_argument("--max-width", type=int, default=2560)
        parser.add_argument("--max-height", type=int, default=2560)
        parser.add_argument("--quality", type=int, default=84)
        parser.add_argument("--responsive", action="store_true", help="Create responsive WebP sidecar files.")
        parser.add_argument(
            "--responsive-widths",
            default=",".join(str(width) for width in RESPONSIVE_WIDTHS),
            help="Comma-separated responsive widths, used with --responsive.",
        )

    def handle(self, *args, **options):
        if options["replace"] and not options["write"]:
            raise CommandError("--replace requires --write.")
        if not 1 <= options["quality"] <= 100:
            raise CommandError("--quality must be between 1 and 100.")
        if options["max_width"] <= 0 or options["max_height"] <= 0:
            raise CommandError("--max-width and --max-height must be positive.")

        roots = self.resolve_roots(options["paths"])
        images = []
        for root in roots:
            images.extend(self.collect_images(root))
        images = sorted(set(images))

        if not images:
            self.stdout.write("No supported images found.")
            return

        responsive_widths = self.parse_responsive_widths(options["responsive_widths"])
        total_before = 0
        total_after = 0
        optimized_count = 0
        sidecar_count = 0

        for image_path in images:
            try:
                result = self.process_image(image_path, options, responsive_widths)
            except Exception as error:
                self.stderr.write(f"ERROR  {image_path}: {error}")
                continue

            total_before += result["source_size"]
            total_after += result["target_size"]
            optimized_count += int(result["optimized"])
            sidecar_count += result["sidecars"]
            marker = "WRITE" if result["optimized"] else ("RESIZE" if result["needs_resize"] else "OK")
            if not options["write"]:
                self.stdout.write(
                    f"{marker:<6} {result['width']}x{result['height']} "
                    f"{self.format_bytes(result['source_size']):>9}  {image_path}"
                )
            elif result["optimized"]:
                self.stdout.write(
                    f"WRITE  {self.format_bytes(result['source_size'])} -> "
                    f"{self.format_bytes(result['target_size'])}  {result['target_path']}"
                )

        if options["write"]:
            self.stdout.write(
                self.style.SUCCESS(
                    f"Optimized {optimized_count}/{len(images)} images. "
                    f"Responsive sidecars: {sidecar_count}. "
                    f"Total: {self.format_bytes(total_before)} -> {self.format_bytes(total_after)}."
                )
            )
        else:
            self.stdout.write(f"\nAudit: {len(images)} images, {self.format_bytes(total_before)} total.")
            self.stdout.write("No files changed. Add --write to optimize.")

    def resolve_roots(self, paths):
        if paths:
            return [Path(path).resolve() for path in paths]
        return [
            Path(settings.MEDIA_ROOT).resolve(),
            Path(settings.BASE_DIR, "main", "static").resolve(),
        ]

    def collect_images(self, root):
        if not root.exists():
            return []
        if root.is_file():
            return [root] if root.suffix.lower() in SUPPORTED_EXTENSIONS else []

        output = []
        for child in root.iterdir():
            if child.is_dir():
                if child.name in SKIPPED_DIRECTORIES:
                    continue
                output.extend(self.collect_images(child))
            elif child.suffix.lower() in SUPPORTED_EXTENSIONS:
                output.append(child)
        return output

    def process_image(self, image_path, options, responsive_widths):
        source_size = image_path.stat().st_size
        output_extension = ".webp" if options["format"] == "webp" else image_path.suffix.lower()
        target_path = image_path.with_suffix(output_extension)

        with Image.open(image_path) as image:
            image = ImageOps.exif_transpose(image)
            width, height = image.size
            needs_resize = width > options["max_width"] or height > options["max_height"]
            output_image = self.resize_inside(image, options["max_width"], options["max_height"])

            if options["write"]:
                self.save_image(output_image, target_path, output_extension, options["quality"])
                if options["replace"] and target_path != image_path:
                    image_path.unlink(missing_ok=True)
                sidecars = self.write_responsive_sidecars(output_image, target_path, responsive_widths, options)
            else:
                sidecars = 0

        target_size = target_path.stat().st_size if options["write"] and target_path.exists() else source_size
        return {
            "width": width,
            "height": height,
            "source_size": source_size,
            "target_size": target_size,
            "target_path": target_path,
            "needs_resize": needs_resize,
            "optimized": options["write"] and (target_path != image_path or needs_resize or target_size < source_size),
            "sidecars": sidecars,
        }

    def resize_inside(self, image, max_width, max_height):
        output = image.copy()
        output.thumbnail((max_width, max_height), Image.Resampling.LANCZOS)
        return output

    def save_image(self, image, path, extension, quality):
        path.parent.mkdir(parents=True, exist_ok=True)
        if extension in {".jpg", ".jpeg"}:
            image = image.convert("RGB")
            image.save(path, "JPEG", quality=quality, optimize=True, progressive=True)
            return
        if extension == ".png":
            image.save(path, "PNG", optimize=True)
            return
        if extension == ".webp":
            image.save(path, "WEBP", quality=quality, method=6)
            return
        raise CommandError(f"Unsupported output extension: {extension}")

    def write_responsive_sidecars(self, image, target_path, widths, options):
        if not options["responsive"]:
            return 0

        created = 0
        source_width, _source_height = image.size
        for width in widths:
            if width >= source_width:
                continue
            sidecar = target_path.with_name(f"{target_path.stem}-{width}w.webp")
            output = image.copy()
            ratio = width / source_width
            height = max(1, round(image.size[1] * ratio))
            output = output.resize((width, height), Image.Resampling.LANCZOS)
            self.save_image(output, sidecar, ".webp", options["quality"])
            created += 1
        return created

    def parse_responsive_widths(self, value):
        widths = []
        for item in str(value or "").split(","):
            item = item.strip()
            if not item:
                continue
            try:
                width = int(item)
            except ValueError as error:
                raise CommandError("--responsive-widths must contain only integers.") from error
            if width <= 0:
                raise CommandError("--responsive-widths values must be positive.")
            widths.append(width)
        return tuple(sorted(set(widths)))

    def format_bytes(self, value):
        if value < 1024:
            return f"{value} B"
        if value < 1024 ** 2:
            return f"{value / 1024:.1f} KB"
        return f"{value / (1024 ** 2):.2f} MB"
